---
title: "Obtain TCGA Dataset"
#author: "Shannon T. Holloway"
#date: "`r Sys.Date()`"
output: html_document
params:
  project: TARGET-NBL
  participant_id_length: 16
  tumor_id_length: 20
  purity_file: "~/Library/CloudStorage/Box-Box/PROJECT 06023: MolClocks/MolClock_Paper_1/1. Analytic Datasets/TCGA/ncomms9971-s2.csv"
  save_data: true
  save_data_as_rds: true
  save_path: "~/Library/CloudStorage/Box-Box/PROJECT 06023: MolClocks/Neuroblastoma_Paper/Datasets/TARGET/"
  GDCdata_dir: "gdc_data"
  rm_GDCdata_dir: false
  clinical_filename: "cohort1.clinical"
  methyl_filename: "cohort1.methyl"
  gene_filename: "cohort1.rnaseq"
  cnv_filename: "cohort1.cnv"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# This package streamlines the data pull from GDC
library(TCGAbiolinks)
library(plyr)
```


#### Query Clinical Data

```{r query_clinical_data}
query_clinical <- GDCquery_clinic(project = params$project, type = "clinical")

participant_subset <- query_clinical$submitter_id
```
`r nrow(query_clinical)` clinical records.

#### Query Methylation Data

```{r query_methyl_data}
methyl.platform <- "Illumina Human Methylation 450"

query_methyl <- GDCquery(project = params$project, 
                         data.category = 'DNA Methylation', 
                         platform = c(methyl.platform),
                         data.type = "Methylation Beta Value",
                         sample.type = c("Primary Tumor"),
                         barcode = participant_subset)

methyl_participant_ids <- substr(getResults(query_methyl, cols = "cases"), 1L, params$participant_id_length)
methyl_tumor_ids <- substr(getResults(query_methyl, cols = "cases"), 1L, params$tumor_id_length)

# We limit our dataset to include only participants with methylation data
has_methyl <- query_clinical$submitter_id %in% methyl_participant_ids
query_clinical$reason_methyl <- !has_methyl
reasons <- c("reason_methyl")
```
Of the `r length(participant_subset)` participants of interest, `r length(unique(methyl_participant_ids))` have methylation data available. There are `r sum(duplicated(methyl_participant_ids))` duplicated participant ids in the methylation dataset and `r sum(duplicated(methyl_tumor_ids))` duplicated tumor ids.

#### Purity

We limit the tumors of interest to those with CPE $\ge 0.6$.

```{r read_purity}
"
#### if there is purity data
purity_data <- read.csv(params$purity_file)

purity_participant_ids <- substr(purity_data$Sample.ID, 1L, params$participant_id_length)
purity_tumor_ids <- substr(purity_data$Sample.ID, 1L, params$tumor_id_length)

# We limit our analyses to tumors with CPE >= 0.60
high_purity <- !is.na(purity_data$CPE) & purity_data$CPE >= 0.60

has_purity <- query_clinical$submitter_id %in% purity_participant_ids[high_purity]
query_clinical$reason_purity <- !has_purity
"
#### if there is no purity data
query_clinical$reason_purity = rep(F, nrow(query_clinical))
####
"
#### if there is purity data
reasons <- c(reasons, 'reason_purity')

common_tumors <- sort(intersect(methyl_tumor_ids, purity_tumor_ids[high_purity]))
"
#### if there is no purity data
common_tumors <- sort(methyl_tumor_ids)
reasons <- c(reasons, 'reason_purity')
####

# common_tumors <- common_tumors[1:10]
```
For the `r length(participant_subset)` participants of interest, there are `r length(common_tumors)` tumors with methylation data and CPE purity $\ge 0.6$. There are `r sum(duplicated(substr(common_tumors, 1L, params$participant_id_length)))` duplicated participant ids in this tumor subset.

#### Participants With More Than One Tumor In Set of Interest

```{r multiple_tumor_per_participant, echo = FALSE}
dups <- duplicated(substr(common_tumors, 1L, params$participant_id_length)) | 
  duplicated(substr(common_tumors, 1L, params$participant_id_length), fromLast = TRUE)
```
`r if(any(dups)) {paste("Participants with multiple tumors: ", paste(unique(substr(common_tumors, 1L, params$participant_id_length)[dups]), collapse = ", "), ".")}`

```{r keep_one_tumor_per_participant, echo = any(dups)}
has_multiple_tumors <- query_clinical$submitter_id %in% {substr(common_tumors, 1L, params$participant_id_length)[dups]}
query_clinical$note_multiple_tumors <- has_multiple_tumors

print(common_tumors[dups])

# We drop the "B" tumor for each participant
# common_tumors <- setdiff(common_tumors, 
#                          paste0(substr(common_tumors[dups], 1L, params$participant_id_length), "-01B"))

# Keep the first tumor (alphabetically)
common_tumors <- common_tumors[! duplicated(substr(common_tumors, 1L, params$participant_id_length))]
```
There are `r length(common_tumors)` tumors with methylation data and CPE purity $\ge 0.6$; `r sum(duplicated(substr(common_tumors, 1L, params$participant_id_length)))` of which are from the same participant.

#### Identify Dataset Participants in Clinical Data

```{r limit_queries_to_common_cases}
# add column to clinical data indicating if participant is kept in our dataset
query_clinical$in_CpG_dataset <- query_clinical$submitter_id %in% substr(common_tumors, 1L, params$participant_id_length)

# ensure that something did not go wrong
code_given <- apply(query_clinical[, reasons], 1L, any)

# There are some tumors that have methylation data but fail purity test or
# pass purity test, but have no methylation data
query_clinical$reason_methyl_and_purity <- !query_clinical$in_CpG_dataset & !code_given
reasons <- c(reasons, "reason_methyl_and_purity")  
code_given <- apply(query_clinical[, reasons], 1L, any)

cat("The purity values for these tumors are\n")
#print(purity_data[purity_participant_ids %in% query_clinical$submitter_id[query_clinical$reason_methyl_and_purity],])
cat("The tumors with methylation data are\n")
print(methyl_tumor_ids[methyl_participant_ids %in% query_clinical$submitter_id[query_clinical$reason_methyl_and_purity]])

colSums(query_clinical[, reasons])

# The analysis dataset will be only the ductal tumors
query_clinical$in_analysis_dataset <- query_clinical$in_CpG_dataset
```

ENSURE THESE TWO NUMBERS AGREE: `r sum(query_clinical$in_CpG_dataset)` and `r sum(!code_given)`.

The total number of cases removed for cause are as follows:
```{r}
colSums(query_clinical[, c(reasons, "note_multiple_tumors")])
```

#### Final Methylation Query

Redefine query to include only participants with tumors in the defined subset.

```{r final_methylation_query}
query_methyl <- GDCquery(project = params$project, 
                         data.category = 'DNA Methylation', 
                         platform = c(methyl.platform),
                         sample.type = c("Primary Tumor"),
                         data.type = "Methylation Beta Value",
                         barcode = unique(substr(common_tumors, 1L, params$participant_id_length)))

```

#### Query Gene Data

Query gene expression data for only those participants in the analysis dataset.

```{r query_gene_data}
query_gene <- GDCquery(project = params$project, 
                       data.category = 'Transcriptome Profiling', 
                       data.type = 'Gene Expression Quantification', 
                       workflow.type = 'STAR - Counts',
                       experimental.strategy = "RNA-Seq",
                       sample.type = c("Primary Tumor"),
                       barcode = query_clinical$submitter_id[query_clinical$in_analysis_dataset])

gene_participant_ids <- substr(getResults(query_gene, cols = "cases"), 1L, params$participant_id_length)
gene_tumor_ids <- substr(getResults(query_gene, cols = "cases"), 1L, params$tumor_id_length)
```
For the `r sum(query_clinical$in_analysis_dataset)` participants in the analysis dataset, `r length(unique(gene_tumor_ids))` gene expression datasets are available. There are `r sum(duplicated(gene_participant_ids))` duplicated participant ids in the gene expression dataset and `r sum(duplicated(gene_tumor_ids))` duplicated tumor ids.

#### Copy Number Variation Data

Query CNV data for only those participants in the analysis dataset.

```{r retrieve_final_dataset}
query_cnv <- GDCquery(project = params$project, 
                      data.category = 'Structural Variation', 
                      platform = c("Affymetrix SNP 6.0"),
                      sample.type = c("Primary Tumor"),
                      data.type = "Copy Number Segment",
                      barcode = query_clinical$submitter_id[query_clinical$in_analysis_dataset])
cnv_cases <- getResults(query_cnv, cols = "cases")

cnv_participant_ids <- substr(cnv_cases, 1L, params$participant_id_length)
cnv_tumor_ids <- substr(cnv_cases, 1L, params$tumor_id_length)

multiple_cnv_measurements <- cnv_cases[duplicated(cnv_tumor_ids) | duplicated(cnv_tumor_ids, fromLast = TRUE)]
multiple_cnv_measurements

# common_tumors <- common_tumors[! duplicated(substr(common_tumors, 1L, params$participant_id_length))]
duplicate_cnv_measurements <- cnv_cases[duplicated(cnv_tumor_ids)]
```
For the `r sum(query_clinical$in_analysis_dataset)` participants in the analysis dataset, `r length(unique(cnv_tumor_ids))` CNV datasets are available. There are `r sum(duplicated(cnv_participant_ids))` duplicated participant ids in the CNV dataset and `r sum(duplicated(cnv_tumor_ids))` duplicated tumor ids.

#### Download and Save

##### Gene Expression Data

```{r download_and_save_gene, eval = params$save_data}
# download and save the gene expression data
GDCdownload(query_gene, directory = paste0(params$save_path, params$GDCdata_dir))
data.seq <- GDCprepare(query_gene, 
                       directory = paste0(params$save_path, params$GDCdata_dir))

# use the raw_counts data for GSEA
raw_counts <- as.data.frame(data.seq@assays@data$unstranded)
colnames(raw_counts) <- rownames(data.seq@colData)

# subset to the specific tumors of interest
# Note this step removes the "B" tumors for patients with more than 1 tumor
raw_counts <- raw_counts[, substr(colnames(raw_counts), 1L, params$tumor_id_length) %in% common_tumors]

# some tumors have multiple gene expression datasets, remove 1
dups <- duplicated(substr(colnames(raw_counts), 1L, params$tumor_id_length))
cat("Tumors with multiple measurements:")
print(substr(colnames(raw_counts), 1L, params$tumor_id_length)[dups])
print(colnames(raw_counts)[dups | duplicated(substr(colnames(raw_counts), 1L, params$tumor_id_length), fromLast = TRUE)])
raw_counts <- raw_counts[, !dups]

# remove Y chromosome genes and store Ensembl IDs and gene names
rm_pary <- grepl("PAR_Y", data.seq@rowRanges$gene_id)
raw_counts$gene_id <- data.seq@rowRanges$gene_id
raw_counts$gene_name <- data.seq@rowRanges$gene_name
raw_counts <- raw_counts[!rm_pary, ]

# Note there are duplicate gene names
dup_rows <- duplicated(data.seq@rowRanges$gene_name) | duplicated(data.seq@rowRanges$gene_name, fromLast = TRUE)
table(data.seq@rowRanges$gene_name[dup_rows])

write.table(raw_counts, file = paste0(params$save_path, params$gene_filename, "_raw_counts.tsv"), 
            sep = "\t")
if (params$save_data_as_rds) saveRDS(raw_counts, 
                                     file = paste0(params$save_path, 
                                                   params$gene_filename, 
                                                   "_raw_counts.rds"))
dim(raw_counts)
rm(raw_counts, data.seq, query_gene)
```

##### Methylation Data

```{r download_and_save_methyl, eval = params$save_data}
# download and save the DNA methylation data
GDCdownload(query_methyl, directory = paste0(params$save_path, params$GDCdata_dir))
data.methyl <- GDCprepare(query_methyl, 
                          directory = paste0(params$save_path, params$GDCdata_dir))

methyl <- data.methyl@assays@data[[1L]]
colnames(methyl) <- rownames(data.methyl@colData)

# subset to the specific tumors in the CpG dataset
methyl <- methyl[, substr(colnames(methyl), 1L, params$tumor_id_length) %in% common_tumors]

# some tumors have multiple methylation measurements, remove 1
dups <- duplicated(substr(colnames(methyl), 1L, params$tumor_id_length))
cat("Tumors with multiple measurements:")
print(substr(colnames(methyl), 1L, params$tumor_id_length)[dups])
print(colnames(methyl)[dups | duplicated(substr(colnames(methyl), 1L, params$tumor_id_length), fromLast = TRUE)])

has_multiple_methyl <- query_clinical$submitter_id %in% substr(colnames(methyl), 1L, params$participant_id_length)[dups]
query_clinical$note_multiple_methyl <- has_multiple_methyl

methyl <- methyl[, !dups]
write.table(methyl, 
            file = paste0(params$save_path, params$methyl_filename, ".tsv"), 
            sep = "\t")
if (params$save_data_as_rds) saveRDS(methyl, 
                                     file = paste0(params$save_path, 
                                                   params$methyl_filename, 
                                                   ".rds"))
dim(methyl)
rm(methyl, data.methyl, query_methyl)
```

##### CNV Data

```{r download_and_save_cnv, eval = params$save_data}
# download and save the DNA methylation data
GDCdownload(query_cnv, directory = paste0(params$save_path, params$GDCdata_dir))
data.cnv <- GDCprepare(query_cnv, 
                       directory = paste0(params$save_path, params$GDCdata_dir))

# subset to the specific tumors
keep_samples <- substr(data.cnv$Sample, 1L, params$tumor_id_length) %in% common_tumors
cnv <- data.cnv[keep_samples, ]

# removing duplicate measurements
cnv <- cnv[!(cnv$Sample %in% duplicate_cnv_measurements), ]
dim(cnv)

write.table(cnv, 
            file = paste0(params$save_path, params$cnv_filename, ".tsv"), 
            sep = "\t")
if (params$save_data_as_rds) saveRDS(cnv, 
                                     file = paste0(params$save_path, 
                                                   params$cnv_filename, 
                                                   ".rds"))
rm(cnv, data.cnv, query_cnv)
```

```{r download_and_save_clinical, eval = params$save_data}
# save the clinical data
# Flatten list columns
query_clinical_flat <- as.data.frame(lapply(query_clinical, function(col) {
  if (is.list(col)) {
    sapply(col, toString)  # Convert list elements to a single string
  } else {
    col
  }
}))

write.table(query_clinical_flat, file = paste0(params$save_path, params$clinical_filename, ".tsv"), sep = "\t")
if (params$save_data_as_rds) saveRDS(query_clinical_flat, 
                                     file = paste0(params$save_path, 
                                                   params$clinical_filename, 
                                                   ".rds"))
```

```{r remove_gdc_data_direction, eval = params$save_data}
if (params$rm_GDCdata_dir) {
  unlink(paste0(params$save_path, params$GDCdata_dir), recursive = TRUE)
}
```

#### Notes

```{r summary_of_notes, eval = params$save_data}
colSums(query_clinical[, c(reasons, 
                           "note_multiple_tumors", 
                           "note_multiple_methyl")])
```

The participants with multiple tumor samples are
```{r multiples_tumor, eval = params$save_data}
query_clinical$submitter_id[query_clinical$note_multiple_tumors]
```
