---
title: "Obtain TARGET Dataset"
date: "`r Sys.Date()`"
output: html_document
params:
  project: TARGET-NBL
  save_data: true
  save_data_as_rds: true
  save_path: "/Users/danielmonyak/Library/CloudStorage/Box-Box/PROJECT 06023: MolClocks/Neuroblastoma_Paper/Datasets/TARGET/"
  GDCdata_dir: "gdc_data"
  rm_GDCdata_dir: False
  clinical_filename: "cohort1.clinical.TEST"
  methyl_filename: "cohort1.methyl.TEST"
  gene_filename: "cohort1.rnaseq.TEST"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# This package streamlines the data pull from GDC
library(TCGAbiolinks)
library(BiocFileCache) # Need to load this directly to set off hook in .Rprofile
                       # deals with issue when calling GDCprepare
library(plyr)

keepN_fields <- function(x, n) {
  fields <- strsplit(x, "-")[[1]]
  paste(fields[1:n], collapse = "-")
}
getParticipantIDs <- function(fullTumorIDs) {
  sapply(fullTumorIDs, function (x) keepN_fields(x, 3), USE.NAMES=F)
}
getTumorIDs <- function(fullTumorIDs) {
  sapply(fullTumorIDs, function (x) keepN_fields(x, 4), USE.NAMES=F)
}
```


#### Query Clinical Data

```{r query_clinical_data}
query_clinical <- GDCquery_clinic(project = params$project, type = "clinical")

participant_subset <- query_clinical$submitter_id
```
`r nrow(query_clinical)` clinical records.

#### Query Methylation Data

```{r query_methyl_data}
methyl.platform <- "Illumina Human Methylation 450"

query_methyl <- GDCquery(project = params$project, 
                         data.category = 'DNA Methylation', 
                         platform = c(methyl.platform),
                         data.type = "Methylation Beta Value",
                         sample.type = c("Primary Tumor"),
                         barcode = participant_subset)

methyl_participant_ids <- getParticipantIDs(getResults(query_methyl, cols = "cases"))
methyl_tumor_ids <- getTumorIDs(getResults(query_methyl, cols = "cases"))

# We limit our dataset to include only participants with methylation data
has_methyl <- query_clinical$submitter_id %in% methyl_participant_ids
query_clinical$reason_methyl <- !has_methyl
reasons <- c("reason_methyl")
```
Of the `r length(participant_subset)` participants of interest, `r length(unique(methyl_participant_ids))` have methylation data available. There are `r sum(duplicated(methyl_participant_ids))` duplicated participant ids in the methylation dataset and `r sum(duplicated(methyl_tumor_ids))` duplicated tumor ids.

#### Participants With More Than One Tumor In Set of Interest

```{r multiple_tumor_per_participant, echo = FALSE}
# methyl_tumor_ids <- methyl_tumor_ids[1:10]
# methyl_tumor_ids <- c("TARGET-30-PAIFXV-01A", "TARGET-30-PAISNS-01A",
#                       "TARGET-30-PAITCI-01A", "TARGET-30-PAIVHE-01A",
#                       "TARGET-30-PAIVMJ-01A", "TARGET-30-PAIVZR-01A",
#                       "TARGET-30-PAIXFZ-01A", "TARGET-30-PAKHAV-01A")

common_tumors <- sort(methyl_tumor_ids)
duplicated.participants <- duplicated(getParticipantIDs(common_tumors)) |
  duplicated(getParticipantIDs(common_tumors), fromLast = TRUE)
duplicated.tumors <- duplicated(common_tumors) | duplicated(common_tumors, fromLast = TRUE)
dups <- duplicated.participants & !duplicated.tumors
```
`r if(any(dups)) {paste("Participants with multiple tumors: ", paste(unique(getParticipantIDs(common_tumors)[dups]), collapse = ", "), ".")}`

```{r keep_one_tumor_per_participant, echo = any(dups)}
has_multiple_tumors <- query_clinical$submitter_id %in% {getParticipantIDs(common_tumors)[dups]}
query_clinical$note_multiple_tumors <- has_multiple_tumors

print(common_tumors[dups])
```
There are `r length(unique(common_tumors))` tumors with methylation data; `r sum(duplicated(getParticipantIDs(common_tumors)))` of which are from the same participant.

#### Identify Dataset Participants in Clinical Data

```{r limit_queries_to_common_cases}
# add column to clinical data indicating if participant is kept in our dataset
query_clinical$in_CpG_dataset <- query_clinical$submitter_id %in% getParticipantIDs(common_tumors)

# ensure that something did not go wrong
# code_given <- apply(query_clinical[, reasons], 1L, any)
code_given <- query_clinical[, 'reason_methyl']

################################################
#################### delete ####################
# There are some tumors that have methylation data but fail purity test or
# pass purity test, but have no methylation data

# query_clinical$reason_methyl_and_purity <- !query_clinical$in_CpG_dataset & !code_given
# reasons <- c(reasons, "reason_methyl_and_purity")  
# code_given <- apply(query_clinical[, reasons], 1L, any)

# cat("The tumors with methylation data are\n")
# print(methyl_tumor_ids[methyl_participant_ids %in% query_clinical$submitter_id[query_clinical$reason_methyl_and_purity]])

################################################
################################################

# colSums(query_clinical[, reasons])
sum(query_clinical[, 'reason_methyl'])

# The analysis dataset
query_clinical$in_analysis_dataset <- query_clinical$in_CpG_dataset
```

ENSURE THESE TWO NUMBERS AGREE: `r sum(query_clinical$in_CpG_dataset)` and `r sum(!code_given)`.

The total number of cases removed for cause are as follows:
```{r}
# colSums(query_clinical[, c(reasons, "note_multiple_tumors")])
colSums(query_clinical[, c('reason_methyl', "note_multiple_tumors")])
```

#### Final Methylation Query

Redefine query to include only participants with tumors in the defined subset.

```{r final_methylation_query}
query_methyl <- GDCquery(project = params$project, 
                         data.category = 'DNA Methylation', 
                         platform = c(methyl.platform),
                         sample.type = c("Primary Tumor"),
                         data.type = "Methylation Beta Value",
                         barcode = unique(getParticipantIDs(common_tumors)))
```

#### Query Gene Data

Query gene expression data for only those participants in the analysis dataset.

```{r query_gene_data}
query_gene <- GDCquery(project = params$project, 
                       data.category = 'Transcriptome Profiling', 
                       data.type = 'Gene Expression Quantification', 
                       workflow.type = 'STAR - Counts',
                       experimental.strategy = "RNA-Seq",
                       sample.type = c("Primary Tumor"),
                       barcode = query_clinical$submitter_id[query_clinical$in_analysis_dataset])

gene_participant_ids <- getParticipantIDs(getResults(query_gene, cols = "cases"))
gene_tumor_ids <- getTumorIDs(getResults(query_gene, cols = "cases"))
```
For the `r sum(query_clinical$in_analysis_dataset)` participants in the analysis dataset, `r length(unique(gene_tumor_ids))` gene expression datasets are available. There are `r sum(duplicated(gene_participant_ids))` duplicated participant ids in the gene expression dataset and `r sum(duplicated(gene_tumor_ids))` duplicated tumor ids.

#### Download and Save

##### Gene Expression Data

```{r download_and_save_gene, eval = params$save_data}
# download and save the gene expression data
GDCdownload(query_gene, directory = paste0(params$save_path, params$GDCdata_dir))
data.seq <- GDCprepare(query_gene, 
                       directory = paste0(params$save_path, params$GDCdata_dir))

# use the raw_counts data for GSEA
raw_counts <- as.data.frame(data.seq@assays@data$unstranded)
colnames(raw_counts) <- rownames(data.seq@colData)

# subset to the specific tumors of interest
# Note this step removes the "B" tumors for patients with more than 1 tumor
raw_counts <- raw_counts[, getTumorIDs(colnames(raw_counts)) %in% common_tumors]

# some tumors have multiple gene expression datasets, remove 1
dups <- duplicated(getTumorIDs(colnames(raw_counts)))
cat("Tumors with multiple measurements:")
print(getTumorIDs(colnames(raw_counts))[dups])
print(colnames(raw_counts)[dups | duplicated(getTumorIDs(colnames(raw_counts)), fromLast = TRUE)])
raw_counts <- raw_counts[, !dups]

# remove Y chromosome genes and store Ensembl IDs and gene names
rm_pary <- grepl("PAR_Y", data.seq@rowRanges$gene_id)
raw_counts$gene_id <- data.seq@rowRanges$gene_id
raw_counts$gene_name <- data.seq@rowRanges$gene_name
raw_counts <- raw_counts[!rm_pary, ]

# Note there are duplicate gene names
dup_rows <- duplicated(data.seq@rowRanges$gene_name) | duplicated(data.seq@rowRanges$gene_name, fromLast = TRUE)
table(data.seq@rowRanges$gene_name[dup_rows])

write.table(raw_counts, file = paste0(params$save_path, params$gene_filename, "_raw_counts.tsv"), 
            sep = "\t")
if (params$save_data_as_rds) saveRDS(raw_counts, 
                                     file = paste0(params$save_path, 
                                                   params$gene_filename, 
                                                   "_raw_counts.rds"))
dim(raw_counts)
rm(raw_counts, data.seq, query_gene)
```

##### Methylation Data

```{r download_and_save_methyl, eval = params$save_data}
# download and save the DNA methylation data
GDCdownload(query_methyl, directory = paste0(params$save_path, params$GDCdata_dir))
data.methyl <- GDCprepare(query_methyl, 
                          directory = paste0(params$save_path, params$GDCdata_dir))

methyl <- data.methyl@assays@data[[1L]]
colnames(methyl) <- rownames(data.methyl@colData)

# subset to the specific tumors in the CpG dataset
methyl <- methyl[, getTumorIDs(colnames(methyl)) %in% common_tumors]

# some tumors have multiple methylation measurements, remove 1
dups <- duplicated(getTumorIDs(colnames(methyl)))
cat("Tumors with multiple measurements:")
print(getTumorIDs(colnames(methyl))[dups])

tumors.dups.methyl <- colnames(methyl)[dups | duplicated(getTumorIDs(colnames(methyl)), fromLast = TRUE)]
print(tumors.dups.methyl)

has_multiple_methyl <- query_clinical$submitter_id %in% getParticipantIDs(colnames(methyl))[dups]
query_clinical$note_multiple_methyl <- has_multiple_methyl

dups.remove <- tumors.dups.methyl[sapply(tumors.dups.methyl, function (x) substr(x, nchar(x)-2, nchar(x))) == '02D']
methyl <- methyl[, !(colnames(methyl) %in% c(dups.remove))]

write.table(methyl, 
            file = paste0(params$save_path, params$methyl_filename, ".tsv"), 
            sep = "\t")
if (params$save_data_as_rds) saveRDS(methyl, 
                                     file = paste0(params$save_path, 
                                                   params$methyl_filename, 
                                                   ".rds"))
dim(methyl)
rm(methyl, data.methyl, query_methyl)
```


##### Retrieve supplementary clinical data

```{r get_ihc_data, eval = params$save_data}
query_clinical_general <- GDCquery(
    project = params$project,
    data.category = "Clinical",
    data.type = "Clinical Supplement",
    data.format = "BCR XML",
    # barcode = query_clinical$submitter_id[query_clinical$in_analysis_dataset]
)

# Download the IHC
GDCdownload(query_clinical_general, directory = paste0(params$save_path, params$GDCdata_dir))
# Import follow up data
clinical_mycnAmp <- GDCprepare_clinic(query_clinical_general, clinical.info = "patient",
                                      directory = paste0(params$save_path, params$GDCdata_dir))
# clinical_ihc <- clinical_ihc[,
#                   c('bcr_patient_barcode',
#                     'breast_carcinoma_estrogen_receptor_status',
#                     'breast_carcinoma_progesterone_receptor_status',
#                     'lab_proc_her2_neu_immunohistochemistry_receptor_status')
#              ]
# query_clinical <- merge(query_clinical, clinical_ihc,
#                                 by.x='submitter_id', by.y='bcr_patient_barcode',
#                                 all.x=T)
# 
# 
# query_clinical$breast_carcinoma_estrogen_receptor_status <-
#   as.character(query_clinical$breast_carcinoma_estrogen_receptor_status)
# query_clinical$breast_carcinoma_progesterone_receptor_status <-
#   as.character(query_clinical$breast_carcinoma_progesterone_receptor_status)
# query_clinical$lab_proc_her2_neu_immunohistochemistry_receptor_status <-
#   as.character(query_clinical$lab_proc_her2_neu_immunohistochemistry_receptor_status)
# 
# # Drop problematic column
# query_clinical <- query_clinical[, names(query_clinical) != 'sites_of_involvement']
```

##### Clinical data

```{r download_and_save_clinical, eval = params$save_data}
# save the clinical data
# Flatten list columns
query_clinical_flat <- as.data.frame(lapply(query_clinical, function(col) {
  if (is.list(col)) {
    sapply(col, toString)  # Convert list elements to a single string
  } else {
    col
  }
}))

write.table(query_clinical_flat, file = paste0(params$save_path, params$clinical_filename, ".tsv"), sep = "\t")
if (params$save_data_as_rds) saveRDS(query_clinical_flat, 
                                     file = paste0(params$save_path, 
                                                   params$clinical_filename, 
                                                   ".rds"))
```

```{r remove_gdc_data_direction, eval = params$save_data}
if (params$rm_GDCdata_dir) {
  unlink(paste0(params$save_path, params$GDCdata_dir), recursive = TRUE)
}
```

#### Notes

```{r summary_of_notes, eval = params$save_data}
# colSums(query_clinical[, c(reasons, "note_multiple_tumors", "note_multiple_methyl")])
colSums(query_clinical[, c('reason_methyl', "note_multiple_tumors", "note_multiple_methyl")])
```

The participants with multiple tumor samples are
```{r multiples_tumor, eval = params$save_data}
query_clinical$submitter_id[query_clinical$note_multiple_tumors]
```
